onnxruntime/core/optimizer/graph_transformer_utils.cc

```diff
@@ -0,0 +1,109 @@
+
+#include "core/optimizer/graph_transformer_utils.h"
+#include "core/optimizer/identity_elimination.h"
+#include "core/optimizer/slice_elimination.h"
+#include "core/optimizer/conv_mul_fusion.h"
+#include "core/optimizer/conv_bn_fusion.h"
+#include "core/optimizer/conv_add_fusion.h"
+#include "core/optimizer/unsqueeze_elimination.h"
+
+namespace onnxruntime {
+
+namespace transformerutils {
+
+Status ValidateTransformerLevel(unsigned int level) {
+  auto t_level = (unsigned int)std::pow(2, level);
+  if (t_level < TransformerLevel::MaxTransformerLevel) {
+    return Status::OK();
+  }
+  return Status(common::ONNXRUNTIME, common::FAIL, "Unsupported level " + level);
+}
+
+void SetTransformerContext(const uint32_t& level, uint32_t& levels_enabled, 
+                           std::vector<TransformerLevel>* all_levels) {
+
+  ORT_ENFORCE(ValidateTransformerLevel(level).IsOK(),
+              "Unsupported transformer level specified", level);
+
+  levels_enabled = TransformerLevel::Default;
+  if (level == 1) {
+    levels_enabled |= TransformerLevel::Level1;
+  } else if (level == 2) {
+    levels_enabled |= TransformerLevel::Level1 | TransformerLevel::Level2;
+  }
+
+  if (all_levels != nullptr) {
+    all_levels->push_back(TransformerLevel::Level2);
+    all_levels->push_back(TransformerLevel::Level1);
+  }
+}
+
+std::vector<std::unique_ptr<RewriteRule>> GenerateRewriteRules(const TransformerLevel& level, 
+                                                               const std::vector<std::string>* custom_list) {
+  std::vector<std::unique_ptr<RewriteRule>> rules;
+  switch (level) {
+    case TransformerLevel::Level1:
+      rules.push_back(std::make_unique<EliminateIdentity>());
+      rules.push_back(std::make_unique<EliminateSlice>());
+      break;
+
+    case TransformerLevel::Level2:
+      break;
+    default:
+      ORT_ENFORCE(false, "Unsupported level" + level);
+  }
+
+  if (custom_list != nullptr && !custom_list->empty()) {
+    std::vector<std::unique_ptr<RewriteRule>> filtered_list;
+    for (const auto& rule_name : *custom_list) {
+      std::for_each(rules.begin(), rules.end(), [&](std::unique_ptr<RewriteRule>& item) {
+        if((item != nullptr) && (item->Name() == rule_name)) {
+          filtered_list.push_back(std::move(item));
+        }
+      });
+    }
+    return filtered_list;
+  }
+
+  return rules;
+}
+
+std::vector<TransformerProviderSet> GenerateTransformers(const TransformerLevel& level, 
```

Comment:
I think the generation of rule-based transformers should be baked into this.
It will become simpler instead of having both GenerateTransformers and GenerateRules. And you will be defining them just in this file.

Reply:
Lets keep this right now. Once we start adding more transformers and rewrite rule we can move the code around if need be. This wont be a fundamental change